### PR TITLE
Feature/issue 1049 ode speedup

### PR DESCRIPTION
**Pull request by wds15**
## Summary

The aim of this PR is to speedup the `coupled_ode_system`. This is achieved by reducing the number of temporaries on the AD stack and saving the `grad` vector which is allocated otherwise. As example I am reporting with the PR the differences before and after this change the evaluation speed of the ODE stress test on this wiki (https://github.com/stan-dev/math/wiki/ODE-stress-test).

On develop I get: 1977ms
On this PR I get: 1754ms

which is >10% speedup. There is some variation in runtimes, but these are considerably smaller than 10%.

As the `coupled_ode_system` is used for the stiff and non-stiff integration this speedup will propagate to all ODE solvers.

## Tests

No new tests as no new behavior is introduced. All existing ODE tests

```
test/unit/math/rev/mat/functor/integrate_ode_adams_prim_test.cpp
test/unit/math/rev/mat/functor/integrate_ode_adams_rev_test.cpp
test/unit/math/rev/mat/functor/integrate_ode_bdf_rev_test.cpp
test/unit/math/rev/mat/functor/integrate_ode_cvodes_grad_rev_test.cpp
test/unit/math/rev/mat/functor/integrate_ode_bdf_prim_test.cpp
test/unit/math/rev/arr/functor/integrate_ode_rk45_grad_test.cpp
test/unit/math/rev/arr/functor/integrate_ode_rk45_tooMuchWork_test.cpp
test/unit/math/rev/arr/functor/integrate_ode_rk45_test.cpp
test/unit/math/prim/arr/functor/integrate_ode_rk45_test.cpp
```

still run OK.

## Side Effects

It's faster.

## Checklist

- [X] Math issue #1049 

- [X] Copyright holder: Sebastian Weber

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [X] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [X] the code is written in idiomatic C++ and changes are documented in the doxygen

- [X] the new changes are tested

### CODE REVIEW

**Comment by syclik** [comment]
@wds15, mind guiding me through the changes?

It would help a lot if you described:
- what you're fixing in the PR (and include it in the issue if not too onerous)
- how you're fixing it (just describing it in the PR)

I tried looking at the code and it's a little dizzying. I'll figure it out eventually, but if you can describe, then I can follow along much quicker.

**Comment by syclik** [comment]
(I'm just trying to cut down on the number of hours I spend on this review; it's important and I want it in sooner rather than later. I just don't have the bandwidth to relearn all about the ode solver for another week or so.)

**Comment by bob-carpenter** [comment]
I can review this one if you want, @syclik.  The logic that's changing is pretty localized manipulation of containers, etc.

**Comment by syclik** [comment]
Sure!

File: stan/math/rev/arr/functor/coupled_ode_system.hpp [hunk 1]
```diff
@@ -499,19 +485,28 @@ struct coupled_ode_system<F, var, var> {
...
-            temp_deriv += z[N_ + N_ * j + k] * grad[k];
+            temp_deriv += z[N_ + N_ * j + k] * y_vars[k].adj();
 
           dz_dt[N_ + i + j * N_] = temp_deriv;
         }
+
+        for (size_t j = 0; j < M_; j++) {
+          double temp_deriv = theta_vars[j].adj();
+          for (size_t k = 0; k < N_; k++)
+            temp_deriv += z[N_ + N_ * N_ + N_ * j + k] * y_vars[k].adj();
```
Comment (bob-carpenter) [inline stan/math/rev/arr/functor/coupled_ode_system.hpp]: How about 
```
size_t n_p_nn = N_ + N_ * N_;
for (size_t j = 0; j < M_; j++) {
  double temp_deriv = theta_vars[j].adj();
  size_t offset = n_p_nn + N_ * j;
  for (size_t k = 0; k < N_; k++) {
    temp_deriv += z[offset + k] * y_vars[k].adj();
    dz_dt[offset + i] = temp_deriv;
  }
}
```

File: stan/math/rev/arr/functor/coupled_ode_system.hpp [hunk 2]
```diff
@@ -499,19 +485,28 @@ struct coupled_ode_system<F, var, var> {
...
+
+          dz_dt[N_ + N_ * N_ + i + j * N_] = temp_deriv;
```
Comment (bob-carpenter) [inline stan/math/rev/arr/functor/coupled_ode_system.hpp]: Why is this `+ i` instead of `+ k`?  It keeps resetting the same value in the loop as written.

**Comment by wds15** [comment]
Thanks all for reviewing. This code is extremley well tested as messing up anything here will show up in all the integrate tests which check it's numerical correctness ... and as we have rk45/adams/bdf there are a few.

I checked your suggestion, Bob, to save the offset and now I think it improves even further:

stress test now: 1649 ms (sometimes I see 1720ms, but consistently a bit less then what I had before)

So I just committed this change and will merge once tests pass. Thanks for spotting this.

**Comment by wds15** [comment]
oh... looks like I need another review for the last changes... @bob-carpenter ?

**Comment by bob-carpenter** [comment]
Will do as soon as GitHub starts responding.  It's timing out my reviews.

**Pull request merged**